sql: index rec with NotVisible indexes

wenyihu6 opened the pull request:
**sql: introduce different index rec types**

Prior to this commit, the logic for index recommendation output string is
handled by checking if the existing index is not nil. This logic can become
really complicated if more index recommendation types are added in the future.
This commit cleans this up by introducing an interface outputCommand to
distinguish different types of index recommendations.

Assists: https://github.com/cockroachdb/cockroach/issues/85477

Release justification: low risk changes to existing functionality

Release note: None

**sql: index rec with NotVisible indexes**

Now that we have introduced a not visible index feature, we should update index
recommendations accordingly. Prior to this commit, the index recommendation
engine treats all indexes as they are visible. If an index already exists in the
table and stores the same explicit column as an index recommendation, this index
recommendation is removed as it is redundant. However, it is possible that this
index is not visible now. This commits fixes this by adding not visible indexes
to the logic. Instead of removing the redundant index directly, it performs a
check to see if the index is not visible.

Fixes: https://github.com/cockroachdb/cockroach/issues/85477

Release justification: this commit does not interfere with existing
functionality and fixes index recommendation to take index visibility into
account.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86564)
<!-- Reviewable:end -->

mgartner reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86564)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @maryliag, @michae2, and @wenyihu6)

___
*[`pkg/sql/opt/indexrec/hypothetical_table.go` line 143 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86564#-NA5EwbMBeA-E0KYT-Ah:-NA5EwbN7ybW1bgvQufZ:b-lm6cb3) ([raw file](https://github.com/cockroachdb/cockroach/blob/754fd792e0afe1f5910aa9989448bc8f37796733/pkg/sql/opt/indexrec/hypothetical_table.go#L143)):*
> ```Go
> // (HypotheticalTable's embedded table).
> func (ht *HypotheticalTable) redundantHypotheticalIndex(index *hypotheticalIndex) bool {
> 	return ht.existingRedundantIndex(index) != nil && !ht.existingRedundantIndex(index).IsNotVisible()
> ```

This does unnecessary work because it calls `existingRedundantIndex` twice. Also, I don't think it will work as you expect - if there are two redundant indexes - one invisible and one visible, it won't find the visible one. I think you need to move the visibility check into the loop in `existingRedundantIndex`. A test for this case would be nice too.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
> 

I see. I will try to find a better way to add this logic.

wenyihu6 committed:
sql: index rec with NotVisible indexes

Now that we have introduced a not visible index feature, we should update index
recommendations accordingly. Prior to this commit, the index recommendation
engine treats all indexes as they are visible. If an index already exists in the
table and stores the same explicit column as an index recommendation, this index
recommendation is removed as it is redundant. However, it is possible that this
index is not visible now. This commits fixes this by adding not visible indexes
to the logic. Instead of removing the redundant index directly, it performs a
check to see if the index is not visible.

Fixes: https://github.com/cockroachdb/cockroach/issues/85477

Release justification: this commit does not interfere with existing
functionality and fixes index recommendation to take index visibility into
account.

Release note: None

wenyihu6 committed:
sql: introduce different index rec types

Prior to this commit, the logic for index recommendation output string is
handled by checking if the existing index is not nil. This logic can become
really complicated if more index recommendation types are added in the future.
This commit cleans this up by introducing an interface outputCommand to
distinguish different types of index recommendations.

Assists: https://github.com/cockroachdb/cockroach/issues/85477

Release justification: low risk changes to existing functionality

Release note: None

wenyihu6 committed:
a

mgartner commented on pkg/sql/opt/indexrec/index_recommendation_set.go:
```diff
@@ -295,39 +322,102 @@ func (ir *indexRecommendation) storingColumns() []tree.Name {
+// outputCommand is an interface that defines methods that an index
+// recommendation type has to implement. This is a helper to print the final
+// index recommendations.
+type outputCommand interface {
```
I think this interface is more complicated than the solution I was working toward in the collab session. I'll try to finish up that PR tomorrow morning.

mgartner replied:
I have a PR up that should simplify things for you: #86754. It adds a `Replacement bool` to the output recommendations to signify the type of recommendation. If you're adding another type, you can use an enum (like [this](https://github.com/cockroachdb/cockroach/blob/6e191d5f91508102ee44e97485ea4358b0c10375/pkg/sql/opt/optbuilder/scope.go#L115-L138)).

wenyihu6 replied:
Thanks for the fix! Looking at the pr now.

wenyihu6 commented:
New PR has been made for this https://github.com/cockroachdb/cockroach/pull/87174.
